File: ML-model/weathergpt-2/app/api/intelligence.py
"""Dedicated intelligence endpoints for the MERN backend (widgets, dashboards)."""
from typing import Any

from fastapi import APIRouter

from app.ml.risk_model import RuleBasedRiskEngine
from app.models.advisory import AdvisoryRequest, AdvisoryResponse
from app.models.anomaly import AnomalyDetectRequest, AnomalyDetectResponse
from app.models.climate import ClimateTrendRequest, ClimateTrendResponse
from app.models.risk import RiskScoreRequest, RiskScoreResponse
from app.services.advisory import generate_advisory
from app.services.anomaly import detect_anomalies
from app.services.climate import analyze_climate
from app.services.risk import calculate_risk, explain_risk

router = APIRouter()


@router.post("/risk/score", response_model=RiskScoreResponse)
def risk_score(request: RiskScoreRequest) -> RiskScoreResponse:
    return calculate_risk(request)


@router.post("/risk/explain")
def risk_explain(request: RiskScoreRequest) -> dict[str, Any]:
    return explain_risk(calculate_risk(request))


@router.post("/advisory", response_model=AdvisoryResponse)
def advisory(request: AdvisoryRequest) -> AdvisoryResponse:
    return generate_advisory(request)


@router.post("/anomaly/detect", response_model=AnomalyDetectResponse)
def anomaly_detect(request: AnomalyDetectRequest) -> AnomalyDetectResponse:
    return detect_anomalies(request)


@router.post("/climate/analyze", response_model=ClimateTrendResponse)
def climate_analyze(request: ClimateTrendRequest) -> ClimateTrendResponse:
    return analyze_climate(request)


@router.get("/model-info")
def model_info() -> dict[str, Any]:
    return {
        "service": "weathergpt-ai",
        "chat": "POST /chat",
        "risk_engine": {
            "model_version": RuleBasedRiskEngine.MODEL_VERSION,
            "endpoint": "POST /risk/score",
            "supported_hazards": [item.value for item in RuleBasedRiskEngine.HAZARD_WEIGHTS],
        },
        "advisory_engine": {
            "endpoint": "POST /advisory",
            "supported_domains": ["agriculture", "travel", "disaster", "urban_general"],
        },
        "anomaly_engine": {
            "endpoint": "POST /anomaly/detect",
            "methodology": "Z-score vs ERA5 same-month reanalysis statistics (live archive)",
        },
        "climate_engine": {
            "endpoint": "POST /climate/analyze",
            "note": "Multi-year ERA5 statistics via Open-Meteo archive; no fabricated climate normals",
        },
        "historical_analysis_engine": {
            "endpoint": "POST /climate/historical-analysis",
            "source": "Open-Meteo Historical Weather API (ERA5 reanalysis)",
            "nwp_verification": "POST /climate/nwp-verification (Previous Runs API, GFS + ECMWF)",
            "note": (
                "Live archive statistics, multi-city comparison, extremes, and 24h-lead "
                "forecast verification. Not official warnings. Not a climate-change attribution."
            ),
        },
        "nwp_hazard_engine": {
            "endpoint": "POST /nwp/hazard-assessment",
            "models": ["GFS", "ECMWF IFS"],
            "access_layer": "Open-Meteo GFS and ECMWF APIs",
            "note": (
                "Rules-based screening of NWP forecasts. Not official IMD warnings. "
                "Does not predict floods or cyclones."
            ),
        },
        "multilingual_support": [
            "en",
            "hi",
            "bn",
            "te",
            "mr",
            "ta",
            "ur",
            "gu",
            "kn",
            "ml",
            "pa",
            "or",
        ],
        "roles": [
            "citizen",
            "farmer",
            "researcher",
            "aviation",
            "marine",
            "climate_analyst",
            "urban_planner",
            "air_quality",
            "flood_disaster",
        ],
        "mern_contract": {
            "required_chat_fields": ["response", "language", "status"],
            "optional_request_fields": ["location", "role", "alert_context"],
            "optional_chat_fields": [
                "intent",
                "role",
                "tool_calls",
                "risk",
                "advisory",
                "follow_up_questions",
                "confidence",
                "latency_ms",
                "tts_hint",
                "disaster_assessment",
                "historical_analysis",
            ],
        },
    }

File: server/routes/intelligence.js
/**
 * intelligence.js
 * Express routes for WeatherGPT 2.0 ML Intelligence Service.
 * Proxies and augments calls to https://weathergpt2-0.onrender.com.
 */
const express = require('express');
const router = express.Router();
const mlClient = require('../services/mlClient');
const { generalLimiter } = require('../middleware/rateLimiter');
const logger = require('../config/logger');

// GET /api/intelligence/health & /model-info
router.get('/health', generalLimiter, async (req, res) => {
  const health = await mlClient.checkHealth();
  res.json({ success: true, data: health });
});

router.get('/model-info', generalLimiter, async (req, res) => {
  try {
    const info = await mlClient.getModelInfo();
    res.json({ success: true, data: info });
  } catch (err) {
    res.status(500).json({ success: false, error: err.message });
  }
});

// POST /api/intelligence/risk-score
router.post('/risk-score', generalLimiter, async (req, res) => {
  try {
    const { location, hazard = 'composite', weather_context, forecast_context, alert_context } = req.body;
    const result = await mlClient.calculateRiskScore({
      location,
      hazard,
      weather_context,
      forecast_context,
      alert_context,
    });
    res.json({ success: true, data: result });
  } catch (err) {
    logger.error(`[API /risk-score] error: ${err.message}`);
    res.status(err.response?.status || 500).json({
      success: false,
      error: err.response?.data?.detail || err.message,
    });
  }
});

// POST /api/intelligence/risk-explain
router.post('/risk-explain', generalLimiter, async (req, res) => {
  try {
    const result = await mlClient.getRiskExplanation(req.body);
    res.json({ success: true, data: result });
  } catch (err) {
    logger.error(`[API /risk-explain] error: ${err.message}`);
    res.status(err.response?.status || 500).json({
      success: false,
      error: err.response?.data?.detail || err.message,
    });
  }
});

// POST /api/intelligence/advisory
router.post('/advisory', generalLimiter, async (req, res) => {
  try {
    const { location, domain = 'agriculture' } = req.body;
    const result = await mlClient.getDomainAdvisory({ location, domain });
    res.json({ success: true, data: result });
  } catch (err) {
    logger.error(`[API /advisory] error: ${err.message}`);
    res.status(err.response?.status || 500).json({
      success: false,
      error: err.response?.data?.detail || err.message,
    });
  }
});

// POST /api/intelligence/anomaly
router.post('/anomaly', generalLimiter, async (req, res) => {
  try {
    const { location } = req.body;
    const result = await mlClient.detectAnomalies({ location });
    res.json({ success: true, data: result });
  } catch (err) {
    logger.error(`[API /anomaly] error: ${err.message}`);
    res.status(err.response?.status || 500).json({
      success: false,
      error: err.response?.data?.detail || err.message,
    });
  }
});

// POST /api/intelligence/climate-analyze
router.post('/climate-analyze', generalLimiter, async (req, res) => {
  try {
    const { location } = req.body;
    const result = await mlClient.analyzeClimate({ location });
    res.json({ success: true, data: result });
  } catch (err) {
    logger.error(`[API /climate-analyze] error: ${err.message}`);
    res.status(err.response?.status || 500).json({
      success: false,
      error: err.response?.data?.detail || err.message,
    });
  }
});

// POST /api/intelligence/nwp-hazard
router.post('/nwp-hazard', generalLimiter, async (req, res) => {
  try {
    const { location, role = 'citizen', hours = 48 } = req.body;
    const result = await mlClient.assessNWPHazards({ location, role, hours });
    res.json({ success: true, data: result });
  } catch (err) {
    logger.error(`[API /nwp-hazard] error: ${err.message}`);
    res.status(err.response?.status || 500).json({
      success: false,
      error: err.response?.data?.detail || err.message,
    });
  }
});

// POST /api/intelligence/historical-analysis
router.post('/historical-analysis', generalLimiter, async (req, res) => {
  try {
    const result = await mlClient.getHistoricalAnalysis(req.body);
    res.json({ success: true, data: result });
  } catch (err) {
    logger.error(`[API /historical-analysis] error: ${err.message}`);
    res.status(err.response?.status || 500).json({
      success: false,
      error: err.response?.data?.detail || err.message,
    });
  }
});

// POST /api/intelligence/nwp-verification
router.post('/nwp-verification', generalLimiter, async (req, res) => {
  try {
    const result = await mlClient.verifyNWPForecast(req.body);
    res.json({ success: true, data: result });
  } catch (err) {
    logger.error(`[API /nwp-verification] error: ${err.message}`);
    res.status(err.response?.status || 500).json({
      success: false,
      error: err.response?.data?.detail || err.message,
    });
  }
});

module.exports = router;
